src/python/pants/core/goals/check.py

```diff
@@ -201,8 +203,33 @@ async def check(
         )
         for request_type in request_types
     )
+
+    environment_names = await MultiGet(
+        Get(
+            EnvironmentName,
+            EnvironmentNameRequest,
+            EnvironmentNameRequest.from_field_set(field_set),
+        )
+        for request in requests
+        for field_set in request.field_sets
+    )
+    request_for_each_environment_name = (
+        # this is actually correct:
+        request
+        for request in requests
+        for _ in request.field_sets
+    )
+    grouped = itertools.groupby(
```

Comment:
`groupby` only works if the collection is already sorted: https://docs.python.org/3/library/itertools.html#itertools.groupby

Would suggest using a defaultdict instead, since it's a bit clearer anyway.

Reply:
the `defaultdict` is indeed a lot clearer in this case. Much improved.